File: J9Ch/src/J_2_Binary_Search/Required_10/_75_Find_Peak_Element_Medium.java
package J_2_Binary_Search.Required_10;
import java.util.*;
/** 75. Find Peak Element
 * Medium
 * Created by tianhuizhu on 6/27/17.
 */
public class _75_Find_Peak_Element_Medium {

    class Solution {
        /**
         * @param A: An integers array.
         * @return: return any of peek positions.
         */
        public int findPeak(int[] A) {
            // write your code here
            int start = 1, end = A.length-2; // 1.答案在之间，2.不会出界
            while(start + 1 <  end) {
                int mid = (start + end) / 2;
                if(A[mid] < A[mid - 1]) {
                    end = mid;
                } else if(A[mid] < A[mid + 1]) {
                    start = mid;
                } else {
                    end = mid;
                }
            }
            if(A[start] < A[end]) {
                return end;
            } else {
                return start;
            }
        }
    }
}

File: J9Ch/src/S_7_Follow_Up/Required_14/_75_Find_Peak_Element_Medium.cpp
/** 75. Find Peak Element
 * Medium*/


class Solution {
public:
    /**
     * @param A: An integers array.
     * @return: return any of peek positions.
     */
    int findPeak(vector<int> A) {
        int l = 1, r = A.size();
        while (l <= r) {
            int mid = (l + r) / 2;
            if (A[mid] > A[mid-1] && A[mid] > A[mid+1])
                return mid;
            if (A[mid] > A[mid-1])
                l = mid + 1;
            else
                r = mid - 1;
        }
        return -1;
    }
};
